File: src/670.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
   public:
    int maximumSwap(int num) {
        int retVal = 0;

        string numStr = to_string(num);
        int numStrSize = numStr.size();
        int maxDigitIndex = -1;
        int swapIdx1 = -1;
        int swapIdx2 = -1;
        for (int i = numStrSize - 1; i >= 0; --i) {
            if ((maxDigitIndex == -1) || (numStr[i] > numStr[maxDigitIndex])) {
                maxDigitIndex = i;
            } else if (numStr[i] < numStr[maxDigitIndex]) {
                swapIdx1 = i;
                swapIdx2 = maxDigitIndex;
            }
        }

        if ((swapIdx1 != -1) && (swapIdx2 != -1)) {
            swap(numStr[swapIdx1], numStr[swapIdx2]);
        }
        retVal = stoi(numStr);

        return retVal;
    }
};

int main(int argc, char** argv) {
    struct subject {
        int num;
    };
    vector<subject> testData{{2736}, {9973}};
    int numberOfTestCase = testData.size();
    /* Example
     *  Input: num = 2736
     *  Output: 7236
     *
     *  Input: num = 9973
     *  Output: 9973
     */

    Solution cSolution;
    int answer;
    for (int i = 0; i < numberOfTestCase; ++i) {
        cout << "Input: num = " << testData[i].num << "\n";

        answer = cSolution.maximumSwap(testData[i].num);
        cout << "Output: " << answer << "\n";

        cout << "\n";
    }

    return EXIT_SUCCESS;
}

File: src/670.py
import logging
import sys
from bisect import bisect_left, bisect_right
from collections import Counter, defaultdict, deque
from functools import cache, lru_cache
from heapq import heapify, heappop, heappush
from itertools import accumulate
from typing import List, Optional


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Solution:
    def maximumSwap(self, num: int) -> int:
        retVal = 0

        numStr = list(str(num))
        numStrSize = len(numStr)
        maxDigitIndex = -1
        swapIdx1 = -1
        swapIdx2 = -1
        for i in range(numStrSize - 1, -1, -1):
            if maxDigitIndex == -1 or numStr[i] > numStr[maxDigitIndex]:
                maxDigitIndex = i
            elif numStr[i] < numStr[maxDigitIndex]:
                swapIdx1 = i
                swapIdx2 = maxDigitIndex

        if swapIdx1 != -1 and swapIdx2 != -1:
            numStr[swapIdx1], numStr[swapIdx2] = numStr[swapIdx2], numStr[swapIdx1]

        retVal = int("".join(numStr))

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for num in [2736, 9973]:
            # /* Example
            #  *  Input: num = 2736
            #  *  Output: 7236
            #  *
            #  *  Input: num = 9973
            #  *  Output: 9973
            #  */
            logging.info("Input: num = %s", num)

            retVal = pSolution.maximumSwap(num)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

    sys.exit(0)
